File: Donggyo/week2/day09/피로도.java
class Solution {
    static boolean[] visited;
    static int max = -1;
    public int solution(int k, int[][] dungeons) {
        int answer = -1;
        visited = new boolean[dungeons.length];
        dfs(k, 0, dungeons, visited);
        answer = max;
        return answer;
    }
    
    static void dfs(int hp, int count, int[][] dungeons, boolean[] visited){
        max = Math.max(count, max);
        for(int i = 0; i<dungeons.length;i++){
            if(visited[i]) continue;
            if(hp<dungeons[i][0]) continue;
            visited[i] = true;
            dfs(hp-dungeons[i][1], count+1, dungeons, visited);
            visited[i] = false;
        }
    }
}

File: Jiho/week2/day09/피로도.js
function solution(k, dungeons) {
  let maxCount = 0;

  function dfs(k, visited, count) {
    maxCount = Math.max(maxCount, count);

    for (let i = 0; i < dungeons.length; i++) {
      if (!visited[i] && k >= dungeons[i][0]) {
        visited[i] = true;
        dfs(k - dungeons[i][1], visited, count + 1);
        visited[i] = false;
      }
    }
  }

  dfs(k, Array(dungeons.length).fill(false), 0);

  return maxCount;
}
